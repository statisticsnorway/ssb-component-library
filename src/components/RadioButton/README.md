RadioButton
========

> A clickable radio selector

### Usage

#### HTML

```html
<div class="ssb-radio">
    <input id="option1" value="option1" type="radio" />
    <label class="radio-label" for="option1">Insert label</label>
</div>
```

#### React

```jsx harmony
<RadioButton
  callback={() => callbackFunction()}
  disabled={false}
  tabIndex="0"
  selected={false}
  value="option1"
>Item 1
</RadioButton>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| callback   | func | Function to trigger when value changes |
| children | node | Label for the checkbox |
| disabled | bool | Toggles disabled state of input field |
| selected | bool | Selected state of checkbox |
| tabIndex | number | Tab index for focus |
| value | Required number or string | Input field value |
