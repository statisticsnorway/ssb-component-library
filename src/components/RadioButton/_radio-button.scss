.ssb-radio {
  @include roboto;
  color: $ssb-dark-6;
  cursor: pointer;
  display: block;
  font-size: 16px;
  min-width: 200px;
  outline: none;
  position: relative;
  text-align: left;
  user-select: none;

  &:hover, &:focus {

    .radio-label {

      &:before {
        border: 2px solid $ssb-green-4;
      }
    }
  }

  input[type=radio] {
    cursor: pointer;
    opacity: 0;
    position: absolute;

    &:focus + .radio-label {

      &:before {
        border: 2px solid $ssb-green-4;
      }
    }

    &:checked + .radio-label {

      &:after {
        background: $ssb-dark-5;
        border-radius: 50%;
        content: '';
        display: block;
        height: 12px;
        left: 4px;
        position: absolute;
        top: 4px;
        transform: rotate(45deg);
        width: 12px;
      }
    }

    &:disabled + .radio-label {
      background: $ssb-white;
      color: $ssb-dark-2;
      cursor: default;

      &:before {
        border: 1px solid $ssb-dark-2;
      }
    }
  }

  .radio-label {
    align-items: center;
    cursor: pointer;
    display: flex;
    margin-bottom: 12px;
    position: relative;

    &:before {
      background: $ssb-white;
      border: 1px solid $ssb-dark-5;
      border-radius: 50%;
      box-sizing: border-box;
      content: '';
      display: block;
      height: 20px;
      margin-right: 12px;
      width: 20px;
    }
  }
}
