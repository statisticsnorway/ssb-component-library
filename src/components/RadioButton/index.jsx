import React from 'react';
import PropTypes from 'prop-types';

const RadioButton = ({
	callback, children, className, disabled, name, selected, tabIndex, value,
}) => (
	<div className={`ssb-radio${className ? ` ${className}` : ''}`}>
		<input
			tabIndex={tabIndex}
			checked={selected}
			id={value}
			name={name}
			disabled={disabled}
			onChange={() => callback(value)}
			type="radio"
			value={value}
		/>
		<label className="radio-label" htmlFor={value}>{children}</label>
	</div>
);

RadioButton.defaultProps = {
	callback: () => {},
	disabled: false,
	tabIndex: 0,
};

RadioButton.propTypes = {
	callback: PropTypes.func,
	children: PropTypes.node.isRequired,
	className: PropTypes.string,
	disabled: PropTypes.bool,
	name: PropTypes.string,
	selected: PropTypes.bool,
	tabIndex: PropTypes.number,
	value: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
	]).isRequired,
};

export default RadioButton;
