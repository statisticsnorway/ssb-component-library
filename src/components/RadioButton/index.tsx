import React, { type ReactNode } from 'react'
import classNames from '../../utils/utils'

interface RadioButtonProps {
  callback?: (value: string | number) => void
  children: ReactNode
  className?: string
  disabled?: boolean
  name?: string
  selected?: boolean
  tabIndex?: number
  value: string | number
  id?: string
}

const RadioButton: React.FC<RadioButtonProps> = ({
  callback = () => {},
  children,
  className,
  disabled = false,
  name,
  selected,
  tabIndex = 0,
  value,
  id,
}) => (
  <div className={classNames('ssb-radio', className)}>
    <input
      tabIndex={tabIndex}
      checked={selected}
      id={id ?? value.toString()}
      name={name}
      disabled={disabled}
      onChange={() => callback(value)}
      type='radio'
      value={value}
    />
    <label className='radio-label' htmlFor={id ?? value.toString()}>
      {children}
    </label>
  </div>
)

export default RadioButton
