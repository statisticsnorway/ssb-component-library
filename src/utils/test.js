import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
