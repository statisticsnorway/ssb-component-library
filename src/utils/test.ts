import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
