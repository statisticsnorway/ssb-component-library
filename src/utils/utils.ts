// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
