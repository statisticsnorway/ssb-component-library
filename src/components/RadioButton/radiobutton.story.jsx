import React from 'react'
import RadioButton from './index'

export default {
  title: 'RadioButton',
}

export const Default = () => (
  <div>
    <RadioButton value='item'>Item</RadioButton>
  </div>
)
