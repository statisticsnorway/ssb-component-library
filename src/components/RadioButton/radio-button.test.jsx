import React from 'react'
import userEvent from '@testing-library/user-event'

import { screen, render } from '../../utils/test'
import RadioButton from './index'

describe('RadioButton component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<RadioButton value='item'>Item</RadioButton>)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Callback is triggered', async () => {
    const callback = jest.fn()
    const user = userEvent.setup()

    render(
      <RadioButton callback={callback} value='item'>
        Item
      </RadioButton>
    )
    const input = screen.getByRole('radio')
    await user.click(input)

    expect(callback).toBeCalled()
  })
})
