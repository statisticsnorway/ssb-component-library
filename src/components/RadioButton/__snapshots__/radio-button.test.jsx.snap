// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-radio"
  >
    <input
      id="item"
      tabindex="0"
      type="radio"
      value="item"
    />
    <label
      class="radio-label"
      for="item"
    >
      Item
    </label>
  </div>
</DocumentFragment>
`;
